src/main/java/spliterators/part1/exercise/RectangleSpliterator.java

```diff
@@ -37,25 +37,28 @@ public OfInt trySplit() {
         if (totalLength < 2){
             return null;
         }
-        final RectangleSpliterator result;
         int lengthLeft = totalLength / 2;
-        if (lengthLeft - (innerLength - startInnerInclusive) <= 0){     // outerLength == 1
-            result = new RectangleSpliterator(array,startOuterInclusive,startOuterInclusive + 1,startInnerInclusive,startInnerInclusive + lengthLeft);
-            startInnerInclusive = startInnerInclusive + lengthLeft;
-            return result;
+        int fullRowsCount = Math.floorDiv(lengthLeft,innerLength);
+        final int remain = lengthLeft % innerLength;
+        int endInnerExclusiveResult = startInnerInclusive + remain;
+        if (endInnerExclusiveResult >= innerLength) {
+            ++fullRowsCount;
+            endInnerExclusiveResult -= innerLength;
         }
-        lengthLeft -= (innerLength - startInnerInclusive);
-        final int fullRowsCount = Math.floorDiv(lengthLeft,innerLength); // outerLength > 1
-        lengthLeft -= fullRowsCount * innerLength;
-        result = new RectangleSpliterator(array,startOuterInclusive,startOuterInclusive + 2 + fullRowsCount,startInnerInclusive, lengthLeft);
-        startInnerInclusive = lengthLeft;
-        startOuterInclusive += 1 + fullRowsCount;
+        final int nextStartOuterInclusive = startOuterInclusive + fullRowsCount;
+        final RectangleSpliterator result = new RectangleSpliterator(array,startOuterInclusive,
+                nextStartOuterInclusive + 1,startInnerInclusive, endInnerExclusiveResult);
+        startInnerInclusive = endInnerExclusiveResult;
+        startOuterInclusive = nextStartOuterInclusive;
         return result;
     }
 
     @Override
     public long estimateSize() {
         int outerLength = endOuterExclusive - startOuterInclusive;
+        if (outerLength <= 0){
+            throw new NegativeArraySizeException("startOuterInclusive == endOuterExclusive");

```

Comment:
Just return `0`.
You can't be sure no one will ever call `estimateSize` on empty spliterator.